crates/turbopack-ecmascript/src/references/require_context.rs

```diff
@@ -0,0 +1,267 @@
+use std::collections::VecDeque;
+
+use anyhow::Result;
+use indexmap::IndexMap;
+use swc_core::{
+    common::DUMMY_SP,
+    ecma::ast::{Expr, KeyValueProp, Lit, ObjectLit, Prop, PropName, PropOrSpread},
+    quote, quote_expr,
+};
+use turbo_tasks::{
+    primitives::{RegexVc, StringVc},
+    Value, ValueToString, ValueToStringVc,
+};
+use turbo_tasks_fs::{DirectoryContent, DirectoryEntry, FileSystemPathVc};
+use turbopack_core::{
+    chunk::{ChunkableAssetReference, ChunkableAssetReferenceVc},
+    issue::{IssueSeverityVc, OptionIssueSourceVc},
+    reference::{AssetReference, AssetReferenceVc},
+    resolve::{
+        origin::{ResolveOrigin, ResolveOriginVc},
+        parse::RequestVc,
+        ResolveResult, ResolveResultVc,
+    },
+};
+
+use crate::{
+    chunk::EcmascriptChunkingContextVc,
+    code_gen::{CodeGenerateable, CodeGeneration, CodeGenerationVc},
+    create_visitor,
+    references::{
+        pattern_mapping::{PatternMappingVc, ResolveType::Cjs},
+        AstPathVc,
+    },
+    resolve::{cjs_resolve, try_to_severity},
+    CodeGenerateableVc,
+};
+
+#[turbo_tasks::value(transparent)]
+pub(crate) struct FlatDirList(Vec<(String, FileSystemPathVc)>);
+
+#[turbo_tasks::function]
+pub(crate) async fn list_dir(
+    dir: FileSystemPathVc,
+    recursive: bool,
+    filter: RegexVc,
+) -> Result<FlatDirListVc> {
+    let root = &*dir.await?;
+    let filter = &*filter.await?;
+
+    let mut queue = VecDeque::with_capacity(1);
+    queue.push_back(dir);
+
+    let mut list = Vec::new();
+
+    while let Some(dir) = queue.pop_front() {
+        let dir_content = dir.read_dir().await?;
+        let entries = match &*dir_content {
+            DirectoryContent::Entries(entries) => entries,
+            DirectoryContent::NotFound => continue,
+        };
+
+        for (_, entry) in entries {
+            match entry {
+                DirectoryEntry::File(path) => {
+                    if let Some(relative_path) = root.get_relative_path_to(&*path.await?) {
+                        if filter.is_match(&relative_path) {
+                            list.push((relative_path, *path));
```

Comment:
It need to apply some resolve options here. e. g. `extensions`.

e.g. a file `./abc/index.ts` should have multiple keys: `./abc/index.ts` `./abc/index` `./abc/` or `./abc`

a package.json with `main` or `exports` would also influence that. (webpack doesn't handle that case)

Reply:
How would the `package.json` influence this?
The API seems to be file path based, so webpack's behavior kinda makes sense to me